lib/Monads/trace/Trace_More_RG.thy

```diff
@@ -0,0 +1,685 @@
+  by (fastforce simp: validI_def)
+
+lemma rg_lift_Pf_pre_conj:
+  assumes P: "\<And>x. \<lbrace>\<lambda>s0 s. Q x s0 s\<rbrace>,\<lbrace>R\<rbrace> m \<lbrace>G\<rbrace>,\<lbrace>P x\<rbrace>"
+  assumes f: "\<And>P. \<lbrace>\<lambda>s0 s. P (g s0 s) \<and> P' s0 s\<rbrace>,\<lbrace>R\<rbrace> m \<lbrace>G\<rbrace>,\<lbrace>\<lambda>_ s0 s. P (f s0 s)\<rbrace>"
+  shows "\<lbrace>\<lambda>s0 s. Q (g s0 s) s0 s \<and> P' s0 s\<rbrace>,\<lbrace>R\<rbrace> m \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. P (f s0 s) rv s0 s\<rbrace>"
+  apply (clarsimp simp: validI_def validI_prefix_closed[OF f] guar_by_rg[OF f])
+  apply (rule use_validI[OF _ P], simp)
+  apply (rule use_validI[OF _ f], simp+)
+  done
+
+lemmas rg_lift_Pf4 = rg_lift_Pf_pre_conj[where P'="\<top>\<top>", simplified]
+lemmas rg_lift_Pf3 = rg_lift_Pf4[where f=f and g=f for f]
+lemmas rg_lift_Pf2 = rg_lift_Pf3[where P="\<lambda>f _. P f" for P]
+lemmas rg_lift_Pf = rg_lift_Pf2[where Q=P and P=P for P]
+
+lemmas rg_lift_Pf3_pre_conj = rg_lift_Pf_pre_conj[where f=f and g=f for f]
+lemmas rg_lift_Pf2_pre_conj = rg_lift_Pf3_pre_conj[where P="\<lambda>f _. P f" for P]
+lemmas rg_lift_Pf_pre_conj' = rg_lift_Pf2_pre_conj[where Q=P and P=P for P]
+
+lemma rg_if_r_and:
+  "\<lbrace>P\<rbrace>,\<lbrace>R'\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>r. if R r then Q r else Q' r\<rbrace>
+   = \<lbrace>P\<rbrace>,\<lbrace>R'\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>r s0 s. (R r \<longrightarrow> Q r s0 s) \<and> (\<not>R r \<longrightarrow> Q' r s0 s)\<rbrace>"
+  by (fastforce simp: validI_def)
+
+lemma rg_convert_imp:
+  "\<lbrakk> \<lbrace>\<lambda>s0 s. \<not> P s0 s\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. \<not> Q s0 s\<rbrace>; \<lbrace>P'\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>S\<rbrace> \<rbrakk>
+   \<Longrightarrow> \<lbrace>\<lambda>s0 s. P s0 s \<longrightarrow> P' s0 s\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. Q s0 s \<longrightarrow> S rv s0 s\<rbrace>"
+  apply (simp only: imp_conv_disj)
+  apply (erule(1) rg_vcg_disj_lift)
+  done
+
+lemma rg_case_option_wpR:
+  "\<lbrakk>\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f None \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>; \<And>x. \<lbrace>P' x\<rbrace>,\<lbrace>R\<rbrace> f (Some x) \<lbrace>G\<rbrace>,\<lbrace>Q' x\<rbrace>,\<lbrace>E\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<lbrace>case_option P P' v\<rbrace>,\<lbrace>R\<rbrace> f v \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv. case v of None \<Rightarrow> Q rv | Some x \<Rightarrow> Q' x rv\<rbrace>,\<lbrace>E\<rbrace>"
+  by (cases v) auto
+
+\<comment> \<open>FIXME MC: delete since we won't use K for rg? If we don't delete, it needs to assume the G
+lemma K_valid[wp]:
+  "\<lbrace>K P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>_. K P\<rbrace>"
+  by (simp add: valid_def)\<close>
```

Comment:
would delete, yes

Reply:
I think this is before I realised that it was safe to leave out the `G` in a lot of these rules, so if we wanted to be complete we could do `⦃K P⦄,⦃R⦄ f -,⦃λ_. K P⦄`. Now that I think about it, I guess that might be useful for postconditions that mix conjuncts that talk about both states with others that only refer to the current state.

Reply:
Happy with that.

Reply:
Whoops, I wasn't thinking properly when I said that yesterday, that rule doesn't make any sense since it would be saying that `f` preserves all `P`s that only talk about the current state. I'll just delete this.